### PR TITLE
refactor code: load/save intermediate files in a concentrated place so it could be atom operation

### PR DESCRIPTION
**Pull request by ansyral**
commit1.
move incrementals code back to engine project since they are mutual-referenced.
commit2.
load/save intermediate files during serialization/deserialization so that once the build failed, buildinfo won't record partial result.
intermediate files could use fixed name so that there is no need to clean up files from last build

### CODE REVIEW

**Comment by qinezh** [comment]
`NonDNX.sln` also need update...

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalUtility.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                return null;
+            }
+            using (var reader = new StreamReader(dependencyFile))
+            {
+                return DependencyGraph.Load(reader);
+            }
+        }
+
+        public static void SaveDependency(string fileName, DependencyGraph dg)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalUtility.cs]: >SaveDependency [](start = 27, length = 14)

change to SaveDependencyGraph to be consistent with LoadDependencyGraph?

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildVersionInfo.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        /// The file link for the XRefMap file(type is <see cref="XRefMap"/>).
+        /// </summary>
+        public string XRefSpecMapFile { get; set; }
+
+        #region Deserialized content
+        /// <summary>
+        /// deserialized dependency graph
+        /// </summary>
+        [JsonIgnore]
+        public DependencyGraph Dependency { get; set; }
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildVersionInfo.cs]: >Dependency [](start = 31, length = 10)

What about move these properties to a new class like BuildVersionInfoModel(maybe other name..), and move the OnDeserializedMethod to BuildVersionInfoModel.From(bvi)?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildVersionInfo.cs]: i don't think it is more clear and we need to track two objects then.

---
In reply to: [75987892](https://github.com/dotnet/docfx/pull/629#discussion_r75987892) [](ancestors = 75987892)

**Comment by superyyrrzz** [comment]
:shipit:

**Comment by qinezh** [comment]
:shipit:

**Comment by ansyral** [comment]
#449

**Pull request merged**